File: frontend/src/features/workflow-admin/models/Workflow.ts
/**
 * Workflow engine types.
 * Field names mirror the backend responses (snake_case) so no mapping layer is
 * needed between the API client and the components.
 */

export interface WorkflowDefinition {
  id: string;
  code: string;
  name: string;
  description: string;
  entity_type: string;
  version: number;
  is_active: boolean;
  created_by: string;
  created_date: string;
  modified_by: string;
  modified_date: string;
}

export interface WorkflowDefinitionListResponse {
  definitions: WorkflowDefinition[];
  total: number;
  skip: number;
  limit: number;
}

export interface WorkflowStatus {
  id: string;
  workflow_definition_id: string;
  code: string;
  name: string;
  is_initial: boolean;
  is_terminal: boolean;
  sequence: number;
}

/**
 * What a transition means for the approval chain.
 *
 * `action_code` is free text, so it cannot tell the engine whether "SIGN_OFF"
 * approves or rejects. This is the designer's explicit declaration, and it is the
 * only signal the approval coordinator reads.
 */
export const WORKFLOW_ACTION_TYPES = [
  'SUBMIT',
  'APPROVE',
  'REJECT',
  'REFER_BACK',
  'CANCEL',
  'CLOSE',
  'ESCALATE',
  'CUSTOM',
] as const;

export type WorkflowActionType = (typeof WORKFLOW_ACTION_TYPES)[number];

/** How each action type affects the approval chain, for display in the builder. */
export const WORKFLOW_ACTION_TYPE_EFFECTS: Record<WorkflowActionType, string> = {
  SUBMIT: 'Opens approval level 1',
  APPROVE: 'Settles this level, opens the next',
  REJECT: 'Settles this level, ends the chain',
  REFER_BACK: 'Ends the chain; a later Submit restarts it at level 1',
  CANCEL: 'Ends the chain',
  CLOSE: 'No effect on approvals',
  ESCALATE: 'No effect on approvals',
  CUSTOM: 'No effect on approvals',
};

export interface WorkflowTransition {
  id: string;
  workflow_definition_id: string;
  from_status_id: string;
  to_status_id: string;
  action_code: string;
  action_type: WorkflowActionType;
  guard_expression: string | null;
  requires_comment: boolean;
  auto_execute: boolean;
  priority: number;
}

/** A definition together with its full state machine, as the builder screen needs. */
export interface WorkflowDefinitionDetail {
  definition: WorkflowDefinition;
  statuses: WorkflowStatus[];
  transitions: WorkflowTransition[];
}

export interface CreateWorkflowDefinitionRequest {
  code: string;
  name: string;
  entity_type: string;
  description: string;
  is_active: boolean;
}

export type UpdateWorkflowDefinitionRequest = Partial<CreateWorkflowDefinitionRequest>;

export interface CreateWorkflowStatusRequest {
  code: string;
  name: string;
  is_initial: boolean;
  is_terminal: boolean;
  sequence: number;
}

export type UpdateWorkflowStatusRequest = Partial<CreateWorkflowStatusRequest>;

export interface CreateWorkflowTransitionRequest {
  from_status_id: string;
  to_status_id: string;
  action_code: string;
  action_type: WorkflowActionType;
  guard_expression?: string | null;
  requires_comment: boolean;
  auto_execute: boolean;
  priority: number;
}

// ─── Runtime ───

export interface WorkflowInstance {
  id: string;
  workflow_definition_id: string;
  definition_code: string;
  definition_name: string;
  entity_type: string;
  entity_id: string;
  current_status_id: string;
  current_status_code: string;
  current_status_name: string;
  is_terminal: boolean;
  initiated_by: string;
  priority: number;
  due_date: string | null;
  started_at: string;
  completed_at: string | null;
  is_completed: boolean;
  /** Approval level currently open; 0 means none. */
  approval_level: number;
  is_awaiting_approval: boolean;
  metadata: Record<string, unknown>;
}

export interface WorkflowInstanceListResponse {
  instances: WorkflowInstance[];
  total: number;
  skip: number;
  limit: number;
}

export interface WorkflowAvailableAction {
  action_code: string;
  action_type: WorkflowActionType;
  to_status_id: string;
  to_status_code: string;
  to_status_name: string;
  requires_comment: boolean;
  is_terminal: boolean;
}

export interface WorkflowHistoryEntry {
  id: string;
  instance_id: string;
  from_status_id: string | null;
  from_status_code: string | null;
  to_status_id: string;
  to_status_code: string | null;
  action_code: string;
  actor_id: string | null;
  actor_username: string;
  comments: string;
  ip_address: string;
  created_at: string;
}

export interface StartWorkflowRequest {
  definition_code: string;
  entity_type: string;
  entity_id: string;
  priority?: number;
  metadata?: Record<string, unknown>;
}

export interface WorkflowActionRequest {
  action_code: string;
  comments: string;
}

File: backend/src/domain/entities/workflow.py
"""Workflow engine domain entities."""

from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any

from src.domain.entities.base_entity import BaseEntity
from src.domain.enums.workflow_enums import WorkflowActionType


@dataclass(kw_only=True)
class WorkflowDefinition(BaseEntity):
    code: str = ""
    name: str = ""
    entity_type: str = ""
    description: str = ""
    version: int = 1
    is_active: bool = True

    def deactivate(self) -> None:
        self.is_active = False


@dataclass(kw_only=True)
class WorkflowStatus(BaseEntity):
    workflow_definition_id: int = 0
    code: str = ""
    name: str = ""
    is_initial: bool = False
    is_terminal: bool = False
    sequence: int = 0


@dataclass(kw_only=True)
class WorkflowTransition(BaseEntity):
    workflow_definition_id: int = 0
    from_status_id: int = 0
    to_status_id: int = 0
    action_code: str = ""
    action_type: WorkflowActionType = WorkflowActionType.CUSTOM
    guard_expression: str | None = None
    requires_comment: bool = False
    auto_execute: bool = False
    priority: int = 0


@dataclass(kw_only=True)
class WorkflowInstance(BaseEntity):
    workflow_definition_id: int = 0
    entity_type: str = ""
    entity_id: int = 0
    current_status_id: int = 0
    initiated_by: int = 0
    priority: int = 0
    due_date: datetime | None = None
    started_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    completed_at: datetime | None = None
    extra_data: dict[str, Any] = field(default_factory=dict)
    approval_level: int = 0

    @property
    def is_completed(self) -> bool:
        return self.completed_at is not None

    @property
    def is_awaiting_approval(self) -> bool:
        return self.approval_level > 0

    def move_to(self, status_id: int, *, is_terminal: bool) -> None:
        self.current_status_id = status_id
        if is_terminal:
            self.completed_at = datetime.now(UTC)


@dataclass(kw_only=True)
class WorkflowHistoryEntry:
    """Immutable record of one executed transition. id is DB-generated."""

    instance_id: int
    to_status_id: int
    action_code: str
    id: int = field(default=0)
    from_status_id: int | None = None
    actor_id: int | None = None
    actor_username: str = ""
    comments: str = ""
    extra_data: dict[str, Any] = field(default_factory=dict)
    ip_address: str = ""
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
